src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1675,6 +1696,9 @@ trait Namers extends MethodSynthesis {
 
   abstract class TypeCompleter extends LazyType {
     val tree: Tree
+    override def forceDirectSuperclasses: Unit = {
+      tree.foreach(t => Option(t.symbol).map(_.maybeInitialize))
```

Comment:
I think that `maybeInitialize` is a weak link in this chain. It is defined as:

```scala
    def maybeInitialize = {
      try   { initialize ; true }
      catch { case _: CyclicReference => debuglog("Hit cycle in maybeInitialize of $this") ; false }
    }
```

The problem is that this doesn't undo any side effects performed by the type completer before it ran into the cycle. The type completer will be run again, and we might observe problems if those side effects aren't idempotent.

`maybeInitialize` is currently only used in `-Ybreak-cycles`, and was described as "sketchy" when it landed: 432f9368011e0fd9e89ca0e18082bfec180baf32